File: dotnet/src/EspBarcode.Protocol/HopFrame.cs
using System.Buffers.Binary;

namespace EspBarcode.Protocol;

public sealed record HopFrameHeader
{
    public byte Major { get; init; } = 2;
    public byte Minor { get; init; }
    public FrameType FrameType { get; init; } = FrameType.Data;
    public byte Flags { get; init; }
    public TrafficClass TrafficClass { get; init; } = TrafficClass.Control;
    public CarrierProfileId ProfileId { get; init; } = CarrierProfileId.Unspecified;
    public ushort RouteId { get; init; }
    public uint LinkSessionId { get; init; }
    public uint LinkMessageId { get; init; }
    public uint LinkCorrelationId { get; init; }
    public ushort FragmentIndex { get; init; }
    public ushort FragmentCount { get; init; } = 1;

    public const int HeaderSize = 32;
    public const int Overhead = 36; // header(32) + crc32 trailer(4)

    public static bool TryEncode(HopFrameHeader header, ReadOnlySpan<byte> payload, out byte[] frame, out CodecError error)
    {
        if (payload.Length > ushort.MaxValue) { frame = []; error = CodecError.PayloadTooLarge; return false; }
        frame = new byte[Overhead + payload.Length];
        var span = frame.AsSpan();
        span[0] = (byte)'E';
        span[1] = (byte)'L';
        span[2] = header.Major;
        span[3] = header.Minor;
        span[4] = (byte)header.FrameType;
        span[5] = header.Flags;
        span[6] = (byte)header.TrafficClass;
        span[7] = (byte)header.ProfileId;
        BinaryPrimitives.WriteUInt16LittleEndian(span[8..], header.RouteId);
        BinaryPrimitives.WriteUInt16LittleEndian(span[10..], HeaderSize);
        BinaryPrimitives.WriteUInt32LittleEndian(span[12..], header.LinkSessionId);
        BinaryPrimitives.WriteUInt32LittleEndian(span[16..], header.LinkMessageId);
        BinaryPrimitives.WriteUInt32LittleEndian(span[20..], header.LinkCorrelationId);
        BinaryPrimitives.WriteUInt16LittleEndian(span[24..], header.FragmentIndex);
        BinaryPrimitives.WriteUInt16LittleEndian(span[26..], header.FragmentCount);
        BinaryPrimitives.WriteUInt16LittleEndian(span[28..], (ushort)payload.Length);
        BinaryPrimitives.WriteUInt16LittleEndian(span[30..], 0);
        payload.CopyTo(span[HeaderSize..]);

        uint crc = Crc32.Compute(span[..(HeaderSize + payload.Length)]);
        BinaryPrimitives.WriteUInt32LittleEndian(span[(HeaderSize + payload.Length)..], crc);
        error = CodecError.None;
        return true;
    }

    public static bool TryDecode(ReadOnlySpan<byte> bytes, out HopFrameHeader header, out byte[] payload, out CodecError error)
    {
        header = new HopFrameHeader();
        payload = [];
        if (bytes.Length < Overhead) { error = CodecError.TooShort; return false; }
        if (bytes[0] != (byte)'E' || bytes[1] != (byte)'L') { error = CodecError.BadMagic; return false; }
        byte major = bytes[2];
        if (major != 2) { error = CodecError.UnsupportedMajorVersion; return false; }
        ushort headerLength = BinaryPrimitives.ReadUInt16LittleEndian(bytes[10..]);
        if (headerLength != HeaderSize) { error = CodecError.BadHeaderLength; return false; }
        ushort payloadLength = BinaryPrimitives.ReadUInt16LittleEndian(bytes[28..]);
        ushort reserved = BinaryPrimitives.ReadUInt16LittleEndian(bytes[30..]);
        if (reserved != 0) { error = CodecError.ReservedFieldNonZero; return false; }
        ushort fragmentIndex = BinaryPrimitives.ReadUInt16LittleEndian(bytes[24..]);
        ushort fragmentCount = BinaryPrimitives.ReadUInt16LittleEndian(bytes[26..]);
        if (fragmentCount == 0 || fragmentIndex >= fragmentCount) { error = CodecError.InvalidFragmentIndex; return false; }

        int rawLength = HeaderSize + payloadLength + 4;
        if (bytes.Length < rawLength) { error = CodecError.TooShort; return false; }

        uint expectedCrc = BinaryPrimitives.ReadUInt32LittleEndian(bytes[(HeaderSize + payloadLength)..]);
        uint actualCrc = Crc32.Compute(bytes[..(HeaderSize + payloadLength)]);
        if (expectedCrc != actualCrc) { error = CodecError.CrcMismatch; return false; }

        header = new HopFrameHeader
        {
            Major = major,
            Minor = bytes[3],
            FrameType = (FrameType)bytes[4],
            Flags = bytes[5],
            TrafficClass = (TrafficClass)bytes[6],
            ProfileId = (CarrierProfileId)bytes[7],
            RouteId = BinaryPrimitives.ReadUInt16LittleEndian(bytes[8..]),
            LinkSessionId = BinaryPrimitives.ReadUInt32LittleEndian(bytes[12..]),
            LinkMessageId = BinaryPrimitives.ReadUInt32LittleEndian(bytes[16..]),
            LinkCorrelationId = BinaryPrimitives.ReadUInt32LittleEndian(bytes[20..]),
            FragmentIndex = fragmentIndex,
            FragmentCount = fragmentCount,
        };
        payload = bytes.Slice(HeaderSize, payloadLength).ToArray();
        error = CodecError.None;
        return true;
    }
}

File: lib/EspLinkCore/src/HopFrame.cpp
#include "HopFrame.h"

#include "Crc32.h"

namespace esplink {

namespace {
void putU16(std::vector<uint8_t>& out, uint16_t v) { out.push_back(uint8_t(v)); out.push_back(uint8_t(v >> 8)); }
void putU32(std::vector<uint8_t>& out, uint32_t v) {
    for (int i = 0; i < 4; ++i) out.push_back(uint8_t(v >> (8 * i)));
}
uint16_t getU16(const uint8_t* p) { return uint16_t(p[0]) | (uint16_t(p[1]) << 8); }
uint32_t getU32(const uint8_t* p) {
    return uint32_t(p[0]) | (uint32_t(p[1]) << 8) | (uint32_t(p[2]) << 16) | (uint32_t(p[3]) << 24);
}
}  // namespace

bool encodeHopFrame(const HopFrameHeader& header, const uint8_t* payload, std::size_t payloadLength,
                    std::vector<uint8_t>& out, CodecError& error) {
    if (payloadLength > 0xFFFFULL) { error = CodecError::PayloadTooLarge; return false; }
    out.clear();
    out.reserve(kHopFrameOverhead + payloadLength);
    out.push_back('E');
    out.push_back('L');
    out.push_back(header.major);
    out.push_back(header.minor);
    out.push_back(static_cast<uint8_t>(header.frameType));
    out.push_back(header.flags);
    out.push_back(static_cast<uint8_t>(header.trafficClass));
    out.push_back(static_cast<uint8_t>(header.profileId));
    putU16(out, header.routeId);
    putU16(out, static_cast<uint16_t>(kHopFrameHeaderSize));
    putU32(out, header.linkSessionId);
    putU32(out, header.linkMessageId);
    putU32(out, header.linkCorrelationId);
    putU16(out, header.fragmentIndex);
    putU16(out, header.fragmentCount);
    putU16(out, static_cast<uint16_t>(payloadLength));
    putU16(out, 0);  // reserved
    out.insert(out.end(), payload, payload + payloadLength);

    const uint32_t crc = crc32(out.data(), out.size());
    putU32(out, crc);
    error = CodecError::None;
    return true;
}

bool decodeHopFrame(const uint8_t* bytes, std::size_t length, HopFrameHeader& header,
                    std::vector<uint8_t>& payload, CodecError& error) {
    if (length < kHopFrameOverhead) { error = CodecError::TooShort; return false; }
    if (bytes[0] != 'E' || bytes[1] != 'L') { error = CodecError::BadMagic; return false; }
    header.major = bytes[2];
    header.minor = bytes[3];
    if (header.major != 2) { error = CodecError::UnsupportedMajorVersion; return false; }
    header.frameType = static_cast<FrameType>(bytes[4]);
    header.flags = bytes[5];
    header.trafficClass = static_cast<TrafficClass>(bytes[6]);
    header.profileId = static_cast<CarrierProfileId>(bytes[7]);
    header.routeId = getU16(bytes + 8);
    const uint16_t headerLength = getU16(bytes + 10);
    if (headerLength != kHopFrameHeaderSize) { error = CodecError::BadHeaderLength; return false; }
    header.linkSessionId = getU32(bytes + 12);
    header.linkMessageId = getU32(bytes + 16);
    header.linkCorrelationId = getU32(bytes + 20);
    header.fragmentIndex = getU16(bytes + 24);
    header.fragmentCount = getU16(bytes + 26);
    const uint16_t payloadLength = getU16(bytes + 28);
    const uint16_t reserved = getU16(bytes + 30);
    if (reserved != 0) { error = CodecError::ReservedFieldNonZero; return false; }
    if (header.fragmentCount == 0 || header.fragmentIndex >= header.fragmentCount) {
        error = CodecError::InvalidFragmentIndex;
        return false;
    }

    const std::size_t rawLength = kHopFrameHeaderSize + std::size_t(payloadLength) + 4U;
    if (length < rawLength) { error = CodecError::TooShort; return false; }

    const uint32_t expectedCrc = getU32(bytes + kHopFrameHeaderSize + payloadLength);
    const uint32_t actualCrc = crc32(bytes, kHopFrameHeaderSize + payloadLength);
    if (expectedCrc != actualCrc) { error = CodecError::CrcMismatch; return false; }

    payload.assign(bytes + kHopFrameHeaderSize, bytes + kHopFrameHeaderSize + payloadLength);
    error = CodecError::None;
    return true;
}

}  // namespace esplink
